service/worker/esanalyzer/analyzer.go

```diff
@@ -0,0 +1,127 @@
+// THE SOFTWARE.
+
+package esanalyzer
+
+import (
+	"context"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber-go/tally"
+	"go.uber.org/cadence/.gen/go/cadence/workflowserviceclient"
+	"go.uber.org/cadence/.gen/go/shared"
+	cclient "go.uber.org/cadence/client"
+	"go.uber.org/cadence/worker"
+
+	"github.com/uber/cadence/client"
+	"github.com/uber/cadence/client/frontend"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cache"
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/dynamicconfig"
+	es "github.com/uber/cadence/common/elasticsearch"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/resource"
+	"github.com/uber/cadence/service/worker/workercommon"
+)
+
+type (
+	// Analyzer is the background sub-system to query ElasticSearch and execute mitigations
+	Analyzer struct {
+		svcClient           workflowserviceclient.Interface
+		frontendClient      frontend.Client
+		clientBean          client.Bean
+		esClient            es.GenericClient
+		logger              log.Logger
+		metricsClient       metrics.Client
+		tallyScope          tally.Scope
+		visibilityIndexName string
+		resource            resource.Resource
+		domainCache         cache.DomainCache
+		config              *Config
+	}
+
+	// Config contains all configs for ElasticSearch Analyzer
+	Config struct {
+		ESAnalyzerLastNDays               dynamicconfig.IntPropertyFn
+		ESAnalyzerLimitToTypes            dynamicconfig.StringPropertyFn
+		ESAnalyzerNumWorkflowsToRefresh   dynamicconfig.IntPropertyFnWithWorkflowTypeFilter
+		ESAnalyzerBufferWaitTimeInSeconds dynamicconfig.IntPropertyFnWithWorkflowTypeFilter
+		ESAnalyzerMinNumWorkflowsForAvg   dynamicconfig.IntPropertyFnWithWorkflowTypeFilter
+	}
+)
+
+// New returns a new instance as daemon
+func New(
+	svcClient workflowserviceclient.Interface,
+	frontendClient frontend.Client,
+	clientBean client.Bean,
+	esClient es.GenericClient,
+	esConfig *config.ElasticSearchConfig,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	tallyScope tally.Scope,
+	resource resource.Resource,
+	domainCache cache.DomainCache,
+	config *Config,
+) *Analyzer {
+	return &Analyzer{
+		svcClient:           svcClient,
+		frontendClient:      frontendClient,
+		clientBean:          clientBean,
+		esClient:            esClient,
+		logger:              logger,
+		metricsClient:       metricsClient,
+		tallyScope:          tallyScope,
+		visibilityIndexName: esConfig.Indices[common.VisibilityAppName],
+		resource:            resource,
+		domainCache:         domainCache,
+		config:              config,
+	}
+}
+
+// Start starts the scanner
+func (a *Analyzer) Start() error {
+	ctx := context.WithValue(context.Background(), analyzerContextKey, a)
+
+	a.StartWorkflow(ctx)
+
+	workerOpts := worker.Options{
+		MetricsScope:              a.tallyScope,
+		BackgroundActivityContext: ctx,
```

Comment:
What's the recommended way to access shared objects? (like Analyzer in this case)

Reply:
register an instance method, basically.
```go
type thatworkflowandactivity struct {
  a Analyzer
}

// plus startup logic like
func startup() {
  t := &thatworkflowandactivity{a: analyzerinstance}
  worker.RegisterWorkflow(t.work)
  worker.RegisterActivity(t.act)
  // though I'll almost always vote for explicit names, ^ that does work.
}

// gives you type safety
func (t *thatworkflowandactivity) work(ctx workflow.Context, ...) {
  t.a.Whatev() // works
  workflow.ExecuteActivity(ctx, t.act, ...) // also works
  workflow.ExecuteActivity(ctx, &thatworkflowandactivity{}.act, ...) // as does an empty value, since only the type matters for reflecting to get the name
}
func (t *thatworkflowandactivity) act(ctx context.Context, ...) {
  t.a.Whatev() // same
}
```